src/useInView.tsx

```diff
@@ -45,56 +45,64 @@ export function useInView({
   fallbackInView,
   onChange,
 }: IntersectionOptions = {}): InViewHookResponse {
+  const [ref, setRef] = React.useState<Element | null>(null);
   const unobserve = React.useRef<Function>();
   const callback = React.useRef<IntersectionOptions['onChange']>();
   const [state, setState] = React.useState<State>({
     inView: !!initialInView,
+    entry: undefined,
   });
-  // Store the onChange callback in a `ref`, so we can access the latest instance inside the `useCallback`.
+
+  // Store the onChange callback in a `ref`, so we can access the latest instance
+  // inside the `useEffect`, but without triggering a rerender.
   callback.current = onChange;
 
-  const setRef = React.useCallback(
-    (node: Element | null) => {
-      if (unobserve.current !== undefined) {
-        unobserve.current();
-        unobserve.current = undefined;
-      }
+  React.useEffect(
+    () => {
+      // Ensure we have node ref, and that we shouldn't skip observing
+      if (skip || !ref) return;
 
-      // Skip creating the observer
-      if (skip) return;
+      unobserve.current = observe(
+        ref,
+        (inView, entry) => {
+          setState({
+            inView,
+            entry,
+          });
+          if (callback.current) callback.current(inView, entry);
 
-      if (node) {
-        unobserve.current = observe(
-          node,
-          (inView, entry) => {
-            setState({ inView, entry });
-            if (callback.current) callback.current(inView, entry);
+          if (entry.isIntersecting && triggerOnce && unobserve.current) {
+            // If it should only trigger once, unobserve the element after it's inView
+            unobserve.current();
+            unobserve.current = undefined;
+          }
+        },
+        {
+          root,
+          rootMargin,
+          threshold,
+          // @ts-ignore
+          trackVisibility,
+          // @ts-ignore
+          delay,
+        },
+        fallbackInView,
+      );
 
-            if (entry.isIntersecting && triggerOnce && unobserve.current) {
-              // If it should only trigger once, unobserve the element after it's inView
-              unobserve.current();
-              unobserve.current = undefined;
-            }
-          },
-          {
-            root,
-            rootMargin,
-            threshold,
-            // @ts-ignore
-            trackVisibility,
-            // @ts-ignore
-            delay,
-          },
-          fallbackInView,
-        );
-      }
+      return () => {
+        if (unobserve.current) {
+          unobserve.current();
+          unobserve.current = undefined;
+        }
+      };
```

Comment:
all looking good to me although one minor thing... it doesn't look like you need the `unobserve` to be a `React.useRef` as you can declare it locally in the effect now:

```jsx
React.useEffect(
  () => {
    if (skip || !ref) return;
    const unobserve = observe(/* .... */);
    return () => unobserve();
  }, 
  [
    /* ... */
  ]
)
```

Reply:
Ah yes. I'll have it inlined.